### PR TITLE
web pages password protected and operational page

### PR DESCRIPTION
**Pull request by knitzsche**
We now require the user to enter a password to use any web page. The password is set by default to wifi-connect when the daemon starts if it is not already set, but we will change the default later to be derived from a device specific term that the user knows, perhaps serial number or mac address. The password is saved as a hash in SNAP_COMMON/hash. You can also set it from the cli with sudo wifi-connect set-hash PASSWORD at any time.

This commit also adds the operational portal page. This displays when the device is connected to external AP. Its only feature now is a "Disconnect from Wifi" button that, when clicked, disconnects, thus causing the daemon to put the device into Management mode. The page needs better design/layout which we can do as follow on work.

Note that the ssids.html page is now named management.html to be consistent with specification terminology, as are all reference to it in code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -114,12 +99,6 @@ sudo  wifi-connect ssid digit
 sudo  wifi-connect passphrase ubuntuubuntuubuntu
```
Comment (knitzsche) [inline README.md]: good points, will fix here

File: README.md [hunk 2]
```diff
@@ -155,15 +144,19 @@ Note: The system trying to open the web page must support Avahi. Android systems
...
-Wifi-connect pauses at daemon startup and at various times to allow state changes to settle. For example:
-
-* On boot and on daemon start, it takes a minute or so to determine the proper state
+Wifi-connect pauses for about a minute at daemon start to allow any external AP connections to complete.
 
 ## Disconnect from wifi
 
+When connected to an external AP, the Operational portal is available on the device IP address (assigned by the external AP). Open it using IP:8080, enter the portal password, and you may then disconnect with the "Disconnect from Wifi" button.
+
+You can alos ssh to the device and:
```
Comment (knitzsche) [inline README.md]: oops, will fix

File: static/templates/management.html [hunk 3]
```diff
@@ -69,6 +76,38 @@
...
+	}
+	function authenticate(){
+            var pw = $('#password').val();
+	    if ( pw.length < 8) {
+		alert("Password must be at least 8 characters long");
+	    } else {
+	        $.ajax({
+	          type: "POST",
+	          url: "/hashit",
+	          data: {Hash: pw}
```
Comment (rmescandon) [inline static/templates/management.html]: I've verified using tcpdump that password can be captured easily as it travels in plain through network:

POST /hashit HTTP/1.1
Host: elsa.local:8080
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:53.0) Gecko/20100101 Firefox/53.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
X-Requested-With: XMLHttpRequest
Referer: http://elsa.local:8080/
Content-Length: 13
Connection: keep-alive

Hash=federico

I'm not a security expert, but I think it is not secure generating the hash in javascript side and, besides, that forces to have a static salt there. So, probably we could consider using ssl. 
We can think about this. Maybe Let's encrypt can be a good solution (though requires internet connection to generate the certificate)

File: try.sh [hunk 4]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (rmescandon) [inline try.sh]: Is this script left on purpose in the project?

File: static/templates/management.html (hunk 3 above)
Comment (knitzsche) [inline static/templates/management.html]: https only seems promising

File: static/templates/management.html [hunk 5]
```diff
@@ -69,6 +76,38 @@
...
+	        }).done(function (hashRet) {
+		    console.log("in ajax done.", hashRet);
+		    hash = JSON.parse(hashRet);
+		    console.log(hash)
+		    if (hash.HashMatch) {
+			showSsids();
+		    } else {
+			alert("Your password does not match, please try again");
```
Comment (knitzsche) [inline static/templates/management.html]: yes, as noted in my PR, the design/layout can/should be improved. Probably in a following PR

File: try.sh (hunk 4 above)
Comment (knitzsche) [inline try.sh]: yes, it is a convenient script that I always use to build and try, and I got tired of recreating it every time I use a fresh image.

File: daemon/daemon.go [hunk 6]
```diff
@@ -105,14 +105,12 @@ func checkWaitApConnect() bool {
 func checkManualMode() bool {
```
Comment (knitzsche) [inline daemon/daemon.go]: actually all or most of daemon functions should have unit tests. I'm refactoring now to do this and will include this in the revised PR

File: utils/utils_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"os"
+	"testing"
+)
+
+func TestHashFile(t *testing.T) {
+	path := "./hash"
+	if _, err := os.Stat(path); !os.IsNotExist(err) {
+		os.Remove(path)
```
Comment (knitzsche) [inline utils/utils_test.go]: right, what the heck was I thinking! ;) fixed

File: static/templates/management.html (hunk 3 above)
Comment (knitzsche) [inline static/templates/management.html]: i've created a backlog card to evaluate https. good catch!

File: static/templates/management.html [hunk 8]
```diff
@@ -12,7 +12,14 @@
 <body>
     <div class="wrapper">
         <div id="main-content" class="inner-wrapper">
-            <div class="row no-border" id="grid">
+	   <div id="login">
+		<p>Password:</p>
+		<input type="text" id="password"/>
```
Comment (knitzsche) [inline static/templates/management.html]: yes, fixing this also in the operational page

File: daemonPkg/daemon.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
+/*
```
Comment (knitzsche) [inline daemonPkg/daemon.go]: I changed the package to "daemon" and the previous daemon dir to service.

File: cmd/main.go [hunk 10]
```diff
@@ -225,8 +237,23 @@ func main() {
 		pw, _ := reader.ReadString('\n')
 		pw = strings.TrimSpace(pw)
 		c.ConnectAp(ssid, pw, ap2device, ssid2ap)
-	case "management-up":
-		http.ListenAndServe(":8081", handler())
+	case "mgmt-up":
```
Comment (knitzsche) [inline cmd/main.go]: "management" "operational" and "set-portal-password" for now.

**Comment by rmescandon** [comment]
LGTM. Please land it and tomorrow morning I'll join to current lp master branch

**Pull request merged**